### ISSUE TITLE
sklearn cross_validate + skorch NeuralNetClassifier results

### ISSUE THREAD

**Issue by angelo5d0** [skorch-dev/skorch#709]
Hello,

Please consider the following code:

`
net = NeuralNetClassifier(
    model,
    max_epochs = 5,
    batch_size = 32,
    optimizer = optim.Adam,
    iterator_train__shuffle = True,
    device = 'cuda',
    verbose = 1
)

result = cross_validate(net,
                        X_train,
                        y_train,
                        cv = 2,
                        scoring = 'accuracy',
                        verbose = 2,
                        n_jobs = 1)

result['test_score']
`

Resulting:

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[CV]  ................................................................
  epoch    train_loss    valid_acc    valid_loss     dur
-------  ------------  -----------  ------------  ------
      1        0.5263       0.7241        0.5608  0.2302
      2        0.5091       0.7186        0.5456  0.2261
      3        0.4840       0.7186        0.5518  0.2214
      4        0.4931       0.7314        0.5300  0.2296
      5        0.4807       **0.7395**        0.5223  0.2321
[CV] ................................................. , total=   1.3s
[CV]  ................................................................
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    1.3s remaining:    0.0s
  epoch    train_loss    valid_acc    valid_loss     dur
-------  ------------  -----------  ------------  ------
      1        0.6133       0.7209        0.5327  0.2415
      2        0.5846       0.7373        0.5215  0.2331
      3        0.5876       0.7377        0.5158  0.2382
      4        0.5771       0.7395        0.5255  0.2631
      5        0.5739       **0.7441**        0.5139  0.2843
[CV] ................................................. , total=   1.4s
[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    2.7s finished

array([0.72817388, 0.78583121])

Why the values in bold differ from the values in the result list?

Thank you.

**Comment by angelo5d0** [comment]
By increasing the verbosity level I got:

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[CV]  ................................................................
  epoch    train_loss    valid_acc    valid_loss     dur
-------  ------------  -----------  ------------  ------
      1        0.5307       0.7236        0.5544  0.2271
      2        0.5064       0.7132        0.5567  0.2145
      3        0.4950       0.7182        0.5421  0.2148
[CV] .................................... , score=0.706, total=   0.8s
[CV]  ................................................................
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.8s remaining:    0.0s
  epoch    train_loss    valid_acc    valid_loss     dur
-------  ------------  -----------  ------------  ------
      1        0.6239       0.7218        0.5337  0.2180
      2        0.5906       0.7323        0.5247  0.2188
      3        0.5785       0.7395        0.5133  0.2145
[CV] .................................... , score=0.785, total=   0.8s
[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    1.6s remaining:    0.0s
[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    1.6s finished

How was the score calculated?

Thanks!

**Comment by BenjaminBossan** [comment]
> Why the values in bold differ from the values in the result list?

I assume that by "result list", you mean the results returned by `cross_validate`.

To understand this, you need to know that skorch performs what we call an _internal_ validation split. So when `net.fit(X_train, y_train)` is called, part of the training data is kept aside. This data is used for the internal validation, and the scores on that data is shown in the printed logs, so e.g. 0.7395 for the validation accuracy.

The score returned by `cross_validate` is calculated based on the train/valid split performed by _sklearn_ itself. So the basic splitting is `data -> cross_validate -> train, valid` and then `train -> skorch -> train_internal, valid_internal`. Therefore, validation is performed on two different datasets, which is the reason why the validation scores differ.

In general, when you use `cross_validate`, `GridSearchCV` etc., I'd recommend to turn off skorch internal validation by passing `train_split=False` to the net. The main reason why you would keep it on is if you need something like early stopping.

**Comment by angelo5d0** [comment]
Thank you very much @BenjaminBossan !

What's the proportion skorch splits the dataset internally?

Regarding the use of skorch's early stopping callback with sklearn's cross_validate, when the patience parameter is reached and the training stops, which validation accuracy is returned by sklearn? the best sklearn's/skorch's validation accuracy found so far or the validation accuracy of the last sklearn/skorch batch?

Thanks!

**Comment by BenjaminBossan** [comment]
> What's the proportion skorch splits the dataset internally?

By default, `train_split=skorch.dataset.CVSplit(5)`, i.e. 20% of data is split off at random (stratified for `NeuralNetClassifier`). If you want to customize this, please refer to the documentation of [`CVSplit`](https://skorch.readthedocs.io/en/stable/user/dataset.html#cvsplit).

> when the patience parameter is reached and the training stops, which validation accuracy is returned by sklearn?

skorch does not automatically restore the parameters from the best epoch. In fact, this would probably be a nice feature to add.

There is an easy way to restore the weights from the best epoch as documented [here](https://skorch.readthedocs.io/en/stable/user/save_load.html#using-callbacks) but this cannot be combined with `cross_validate` unfortunately.
